src/modules/flow/http-server/http-server.c

```diff
@@ -713,6 +714,167 @@ float_process_cb(struct http_data *mdata, const struct sol_flow_packet *packet)
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "green"))
+        STRTOL(rgb.green);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "blue"))
+        STRTOL(rgb.blue);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "red_max"))
+        STRTOL(rgb.red_max);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "green_max"))
+        STRTOL(rgb.green_max);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "blue_max"))
+        STRTOL(rgb.blue_max);
+    else
+        return -EINVAL;
+
+    return 0;
+}
+
+static int
+rgb_process_cb(struct http_data *mdata, const struct sol_flow_packet *packet)
+{
+    return sol_flow_packet_get_rgb(packet, &mdata->value.rgb);
+}
+
+static int
+rgb_open(struct sol_flow_node *node, void *data,
+    const struct sol_flow_node_options *options)
+{
+    int r;
+    struct http_data *mdata = data;
+    struct sol_flow_node_type_http_server_rgb_options *opts =
+        (struct sol_flow_node_type_http_server_rgb_options *)options;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options,
+        SOL_FLOW_NODE_TYPE_HTTP_SERVER_RGB_OPTIONS_API_VERSION,
+        -EINVAL);
+
+    r = common_open(node, mdata, opts->path, opts->allowed_methods, opts->port);
+    SOL_INT_CHECK(r, < 0, r);
+
+    mdata->value.rgb = opts->value;
+    return 0;
+}
+
+static int
+direction_vector_post_cb(struct http_data *mdata, struct sol_flow_node *node,
+    struct sol_http_param_value *value)
+{
+    if (sol_str_slice_str_eq(value->value.key_value.key, "x"))
+        STRTOD(dir_vector.x);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "y"))
+        STRTOD(dir_vector.y);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "z"))
+        STRTOD(dir_vector.z);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "min"))
+        STRTOD(dir_vector.min);
+    else if (sol_str_slice_str_eq(value->value.key_value.key, "max"))
+        STRTOD(dir_vector.min);
+    else
+        return -EINVAL;
+
+    return 0;
+}
+
+static int
+direction_vector_response_cb(struct http_data *mdata,
+    struct sol_buffer *content, bool json)
+{
+    int r;
+
+    SOL_BUFFER_DECLARE_STATIC(x, DOUBLE_STRING_LEN);
+    SOL_BUFFER_DECLARE_STATIC(y, DOUBLE_STRING_LEN);
+    SOL_BUFFER_DECLARE_STATIC(z, DOUBLE_STRING_LEN);
+    SOL_BUFFER_DECLARE_STATIC(min, DOUBLE_STRING_LEN);
+    SOL_BUFFER_DECLARE_STATIC(max, DOUBLE_STRING_LEN);
+
+    r = sol_json_double_to_str(mdata->value.dir_vector.x, &x);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sol_json_double_to_str(mdata->value.dir_vector.y, &y);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sol_json_double_to_str(mdata->value.dir_vector.z, &z);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sol_json_double_to_str(mdata->value.dir_vector.min, &min);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sol_json_double_to_str(mdata->value.dir_vector.max, &max);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return sol_buffer_append_printf(content,
```

Comment:
like I said for `rgb`, we should see what to do with `json`.

BTW, I've just noticed that you create 6 buffers, format them to string, then create the final json... which is a waste! The `content` is already a buffer, you can progressively encode to it and save intermediate buffers and copies.